### PR TITLE
Multi displays support

### PR DESCRIPTION
**Pull request by adglkh**
## Done

This is the initial multi-display support in the JS SDK, which enabling the
WebRTC manager to handle multiple simultaneous video tracks, touch 
tracking and input event routing per-display.

The targetElement (video container) is now optional, preserved for backward 
compatibility in legacy single-display mode. 

Two new callbacks
- videoTrackAdded
- videoTrackRemoved

are added in the JS SDK to support dynamical display attachment(multi-display mode)
via attachDisplay() API.

## QA

1. Enable newly added tests cases pass
2. For the real e2e testing, please try out the [webrtc_multiple_display](https://github.com/adglkh/ams/tree/webrtc_multiple_display), which has this JS SDK mirrored in the dashboard UI
    - Launch an instance multi displays
          $ amc launch resolute:android16-cf:amd64 --enable-streaming --devmode -c 4 -m 8GB --name test-multi-display \\
             --display=width=1280,height=720,fps=60 \\
             --display=width=480,height=640,dpi=120 \\
             --display=width=800,height=600,fps=30
      Once the instance is up and running, ensure each display are present on the UI properly
    - In the meanwhile, present different apps in different displays within the instance
        $ amc exec <instance_id> --  adb shell am start-activity -W -n com.android.dialer/.main.impl.MainActivity --display 2
        $ amc exec <instance_id> --  adb shell am start-activity -W -n com.android.gallery3d/com.android.gallery3d.app.Gallery --display 3 

## JIRA / Launchpad bug

Fixes https://warthogs.atlassian.net/browse/AC-4542

## Documentation

Does this impact the team's internal or public documentation? If yes, has the relevant documentation been updated?

A: yes, in the 1.31.0 release announcement.

### CODE REVIEW

File: js/tests/unit/multi-display.test.js [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+    stream.attachDisplay(0, "main-container");
+
+    const video = document.getElementById(stream._videoID);
+    expect(video).not.toBeNull();
+    expect(video.srcObject).toBe(fakeStream);
+    expect(stream._containerIDs[0]).toBe("main-container");
+    expect(stream._pendingVideoTracks[0]).toBeUndefined();
+  });
+
+  test("attachDisplay(1, containerId) injects secondary video and registers input handlers", () => {
```
Comment (lorumic) [inline js/tests/unit/multi-display.test.js:282]: This doesn't seem to test whether the input handlers are registered, or am I missing something?

```suggestion
  test("attachDisplay(1, containerId) injects secondary video", () => {
```

File: js/tests/unit/multi-display.test.js [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+      }),
+    );
+
+    expect(mockSend).toHaveBeenCalledWith(
+      "input::mouse-button",
+      expect.objectContaining({ display_id: 1 }),
+    );
+  });
+
+  test("events on display 0 container include display_id: 0", () => {
```
Comment (lorumic) [inline js/tests/unit/multi-display.test.js]: Do we need this test? Seems redundant as it duplicates the test above entirely. Also code duplication should be avoided - if we really need 2 tests, we should have a reusable function which accepts what changes in the 2 tests as parameters, and reuse it in the 2 occurrences.

File: js/tests/unit/multi-display.test.js [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+
+  test("resets _containerIDs and _displayStates to initial values", () => {
+    const stream = makeStreamWithTarget("main-container");
+    stream._containerIDs = { 0: "main-container", 1: "cell-1" };
+
+    stream._removeMedia();
+
+    expect(stream._containerIDs).toEqual({ 0: "main-container" });
+    // Should have one display state entry (for display 0).
+    expect(0 in stream._displayStates).toBe(true);
```
Comment (lorumic) [inline js/tests/unit/multi-display.test.js:492]: Should we also verify the absence of the other display state entry?

```suggestion
    expect(0 in stream._displayStates).toBe(true);
    expect(1 in stream._displayStates).toBe(false);
```

File: js/tests/unit/multi-display.test.js [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+    const stream = makeStream({
+      targetElement: "main-container",
+      callbacks: { videoTrackRemoved: (i) => removed.push(i) },
+    });
+
+    stream._containerIDs = { 0: "main-container", 1: "cell-1", 2: "cell-2" };
+    stream._removeMedia();
+
+    expect(removed).toEqual([2, 1, 0]);
+  });
```
Comment (lorumic) [inline js/tests/unit/multi-display.test.js]: This fully duplicates the test at line 436. Can we remove it?

```suggestion
```

File: js/anbox-stream-sdk.js [hunk 5]
```diff
@@ -86,7 +86,12 @@ class AnboxStream {
    * displays its video & audio feed in an HTML5 player
    * @param options: {object}
    * @param options.connector {object} WebRTC Stream connector.
-   * @param options.targetElement {string} ID of the DOM element to attach the video to.
+   * @param [options.targetElement] {string} ID of the DOM element to attach the primary (display 0)
+   *   video to. When provided the SDK operates in legacy single-display mode: display 0 is
+   *   injected automatically, and the `videoTrackAdded` callback fires for every display that arrives
+   *   (including display 0). When omitted the SDK operates in fully dynamic mode: the caller MUST
+   *   call `attachDisplay(displayId, containerId)` for every display, including display 0, in
+   *   response to the `videoTrackAdded` callback. Array values are not supported.
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: We might not need this specification. The type `{string}` should remove any doubt.

```suggestion
   *   response to the `videoTrackAdded` callback.
```

File: js/anbox-stream-sdk.js [hunk 6]
```diff
@@ -316,6 +347,130 @@ class AnboxStream {
...
+        { once: true },
+      );
+
+      // Recompute when the container cell resizes.
+      if (typeof ResizeObserver !== "undefined") {
+        const ro = new ResizeObserver(() => {
+          if (video.videoWidth > 0 && video.videoHeight > 0) computeDims();
+        });
+        ro.observe(container);
+      }
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: Can we try to reuse the code from the if block above (lines 417-437) instead of duplicating it?

File: js/anbox-stream-sdk.js [hunk 7]
```diff
@@ -809,24 +970,25 @@ class AnboxStream {
...
 
-    if (this._nullOrUndef(options.targetElement)) {
-      throw newError(
-        "missing targetElement parameter",
-        ANBOX_STREAM_SDK_ERROR_INVALID_ARGUMENT,
-      );
+    if (!this._nullOrUndef(options.targetElement)) {
+      if (typeof options.targetElement !== "string") {
+        throw newError(
+          "targetElement must be a string ID (array values are not supported)",
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: Again, I think we don't need to specify explicitly that array values are not supported. The user might have passed any non-string type, and would be puzzled to read "array values are not supported" in such case.

```suggestion
          "targetElement must be a string ID",
```

File: js/anbox-stream-sdk.js [hunk 8]
```diff
@@ -809,24 +970,25 @@ class AnboxStream {
...
+          ANBOX_STREAM_SDK_ERROR_INVALID_ARGUMENT,
+        );
+      }
+      const container = document.getElementById(options.targetElement);
+      if (container === null) {
+        throw newError(
+          `target element "${options.targetElement}" does not exist`,
+          ANBOX_STREAM_SDK_ERROR_INVALID_ARGUMENT,
+        );
+      } else if (container.clientWidth == 0 || container.clientHeight == 0)
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: Nitpick - can we use strict equality comparisons consistently?

```suggestion
      } else if (container.clientWidth === 0 || container.clientHeight === 0)
```

File: js/anbox-stream-sdk.js [hunk 9]
```diff
@@ -970,16 +1176,134 @@ class AnboxStream {
...
+    this._multiDisplayActive = true;
+
+    // Wrap the primary video in a dedicated cell div so it becomes one
+    // equal grid item alongside the additional display cells.
+    const cell = document.createElement("div");
+    cell.id = `${this._videoID}-cell-0`;
+    cell.className = "anbox-stream-cell";
+    cell.style.position = "relative";
+    cell.style.overflow = "hidden";
+    cell.style.background = "#000";
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: Do we really need to enforce a black background for the cell? Shouldn't this be a cosmetic user's choice?

File: js/anbox-stream-sdk.js [hunk 10]
```diff
@@ -1473,74 +1883,134 @@ class AnboxStream {
...
+        playerOffsetLeft: contentOffsetX,
+        playerOffsetTop: contentOffsetY,
+      };
+    } else {
+      state.dimensions.scalePercentage = scale;
+      state.dimensions.playerWidth = renderedWidth;
+      state.dimensions.playerHeight = renderedHeight;
+      state.dimensions.playerOffsetLeft = contentOffsetX;
+      state.dimensions.playerOffsetTop = contentOffsetY;
+    }
```
Comment (lorumic) [inline js/anbox-stream-sdk.js:1960]: We can reuse some code here:

```suggestion
    const newValues = {
      scalePercentage: scale,
      playerWidth: renderedWidth,
      playerHeight: renderedHeight,
      playerOffsetLeft: contentOffsetX,
      playerOffsetTop: contentOffsetY,
    };
    if (!state.dimensions) {
      state.dimensions = {
        videoWidth: video.videoWidth,
        videoHeight: video.videoHeight,
        ...newValues,
      };
    } else {
      state.dimensions = { ...state.dimensions, ...newValues };
    }
```

File: js/anbox-stream-sdk.js [hunk 11]
```diff
@@ -1473,74 +1883,134 @@ class AnboxStream {
...
    * Returns true if a pointer event (move or click) was emitted outside the video
-   * boundaries
+   * boundaries of the given display.
+   * @param {object} event
+   * @param {number} displayId
    * @returns {boolean}
    * @private
    */
-  _isPointerEventOutOfBounds(event) {
+  _isPointerEventOutOfBounds(event, displayId) {
```
Comment (lorumic) [inline js/anbox-stream-sdk.js:1874]: I see that many methods are adding the `displayId` parameter now, but not all have a default value for it. Should we make it consistent and always default to 0 for full backward compatibility?

```suggestion
  _isPointerEventOutOfBounds(event, displayId = 0) {
```

File: js/anbox-stream-sdk.js [hunk 12]
```diff
@@ -1473,74 +1883,134 @@ class AnboxStream {
...
   /**
-   * onPointerEvent is called when a mouse or touch input is fired
-   * @param pointerEvent
+   * onPointerEvent is called when a mouse or touch input is fired.
+   * @param pointerEvent {PointerEvent}
+   * @param {number} [displayId=0] Index of the display that received the event.
    * @private
    */
-  _onPointerEvent(pointerEvent) {
+  _onPointerEvent(pointerEvent, displayId = 0) {
```
Comment (lorumic) [inline js/anbox-stream-sdk.js:1998]: Here, unlike the other methods above, we have a default value for `displayId` :)

File: js/anbox-stream-sdk.js [hunk 13]
```diff
@@ -1551,6 +2021,9 @@ class AnboxStream {
     )
       return;
 
+    const state = this._displayStates[displayId];
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: Maybe, just to be a bit clearer and improve code readability some tens of lines below:

```suggestion
    const displayState = this._displayStates[displayId];
```

File: js/anbox-stream-sdk.js [hunk 14]
```diff
@@ -3284,16 +3788,40 @@ class AnboxWebRTCManager {
...
+      // MediaStream object for every track. A per-track MediaStream guarantees
+      // each video element renders its own display independently.
+      const singleTrackStream = new MediaStream([event.track]);
+      this._videoStreams[displayId] = singleTrackStream;
+
+      if (displayId === 0) {
+        this._videoStream = singleTrackStream;
+        if (event.streams[0]) event.streams[0].onremovetrack = this._onClose;
+      } else {
+        // External displays and notify AnboxStream to wire a new video element.
```
Comment (lorumic) [inline js/anbox-stream-sdk.js]: I don't understand what this comment means (especially the first part "External displays"?) - can we make it a bit clearer please?

**Review by lorumic — COMMENTED** [review]
Thank you @adglkh, this is really a monumental work. I tested its integration with the dashboard and everything worked really well, so I'm just leaving some minor comments, nitpicks, and ideas to optimise the code and avoid duplication. Excellent job overall! 💯

File: js/anbox-stream-sdk.js [hunk 15]
```diff
@@ -3284,16 +3788,40 @@ class AnboxWebRTCManager {
...
-      this._videoStream.onremovetrack = this._onClose;
+      // The server assigns each video track the label "video_N" (where N is the
+      // display id) when adding video transceiver. On client side, parsing it
+      // here from MediaStreamTrack.id gives us the correct display id.
+      let displayId;
+      const m = event.track.id.match(/^video_(\d+)$/);
+      if (m) {
+        displayId = parseInt(m[1], 10);
+      } else {
+        console.error(`failed to paser display id: ${event.track.id}`);
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Thanks, fixed that!
When checking this piece of code, I found a regression, which the older images use the legacy track name "video" and we need to fix this.

File: js/anbox-stream-sdk.js [hunk 16]
```diff
@@ -3284,16 +3788,40 @@ class AnboxWebRTCManager {
...
+      // the same stream_id on the server side so event.streams[0] is the same
+      // MediaStream object for every track. A per-track MediaStream guarantees
+      // each video element renders its own display independently.
+      const singleTrackStream = new MediaStream([event.track]);
+      this._videoStreams[displayId] = singleTrackStream;
+
+      if (displayId === 0) {
+        this._videoStream = singleTrackStream;
+        if (event.streams[0]) event.streams[0].onremovetrack = this._onClose;
+      } else {
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:3807]: +1, for the primary display, when event.track.onended is fired, we probably need to trigger _onClose while for the external displays, we need to trigger videoTrackRemoved(displayId) instead .

[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:3807]: Fixed it

File: js/anbox-stream-sdk.js [hunk 17]
```diff
@@ -970,16 +1176,134 @@ class AnboxStream {
...
+      primaryVideo.style.maxHeight = "";
+      cell.appendChild(primaryVideo);
     }
+    container.insertBefore(cell, container.firstChild);
+
+    // Move the display 0 input zone from the outer container to cell-0 so
+    // pointer events are scoped to the actual video area.
+    this._unregisterDisplayControls(0);
+    this._containerIDs[0] = cell.id;
+    this._registerInputHandlers(0, cell);
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:1206]: +1, thanks for pointing this out.
Fixed it.

File: js/anbox-stream-sdk.js [hunk 18]
```diff
@@ -1834,7 +2318,7 @@ class AnboxStream {
    * @private
    */
   _convertTouchInput(x, y) {
-    const dim = this._dimensions;
+    const dim = this._displayStates[0].dimensions;
     if (!dim) {
       throw newError("sdk is not ready yet", ANBOX_STREAM_SDK_ERROR_INTERNAL);
     }
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js]: +1, Fixed that.
Thanks

File: js/anbox-stream-sdk.js [hunk 19]
```diff
@@ -1473,74 +1883,134 @@ class AnboxStream {
...
    */
-  _translateLocalCoordsToRemoteCoords(event) {
+  _translateLocalCoordsToRemoteCoords(event, displayId) {
+    const dim = this._displayStates[displayId].dimensions;
+
     if (event.pointerType === "touch") {
       const pos = this._convertTouchInput(event.clientX, event.clientY);
       event.clientX = pos.x;
       event.clientY = pos.y;
     }
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:1983]: See below.

File: js/anbox-stream-sdk.js [hunk 20]
```diff
@@ -970,16 +1176,134 @@ class AnboxStream {
...
+      );
+      if (el) el.remove();
+    }
+
+    // Remove primary video and audio elements.
     const video = document.getElementById(this._videoID);
     const audio = document.getElementById(this._audioID);
-
     if (video) video.remove();
     if (audio) audio.remove();
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:1265]: The underlyig issue was that the stale callback attached to the video container could be still fired after the session is torn down.
Make sense to me.
Fixed it

File: js/anbox-stream-sdk.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js]: +1, please check the last commit.

File: js/anbox-stream-sdk.js [hunk 21]
```diff
@@ -316,6 +347,130 @@ class AnboxStream {
...
+        { once: true },
+      );
+
+      // Recompute when the container cell resizes.
+      if (typeof ResizeObserver !== "undefined") {
+        const ro = new ResizeObserver(() => {
+          if (video.videoWidth > 0 && video.videoHeight > 0) computeDims();
+        });
+        ro.observe(container);
+      }
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Same below

File: js/anbox-stream-sdk.js [hunk 22]
```diff
@@ -809,24 +970,25 @@ class AnboxStream {
...
-        "missing targetElement parameter",
-        ANBOX_STREAM_SDK_ERROR_INVALID_ARGUMENT,
-      );
+    if (!this._nullOrUndef(options.targetElement)) {
+      if (typeof options.targetElement !== "string") {
+        throw newError(
+          "targetElement must be a string ID (array values are not supported)",
+          ANBOX_STREAM_SDK_ERROR_INVALID_ARGUMENT,
+        );
+      }
```
[the remark this answers is not in the document]
Comment (adglkh) [reply js/anbox-stream-sdk.js:956]: Make sense to me.
Thanks

File: js/tests/unit/multi-display.test.js (hunk 1 above)
Comment (adglkh) [reply js/tests/unit/multi-display.test.js:282]: Oh, we need to improve the test case to cover that.

Comment (adglkh) [reply js/tests/unit/multi-display.test.js:282]: Done

File: js/tests/unit/multi-display.test.js (hunk 2 above)
Comment (adglkh) [reply js/tests/unit/multi-display.test.js]: Hmm, they're not redundant. Both of them covers the scenario where input events are routed to different displays (primary vs additional). But your point is valid that we should avoid code repetition.
Let me combine both

File: js/tests/unit/multi-display.test.js (hunk 3 above)
Comment (adglkh) [reply js/tests/unit/multi-display.test.js:492]: +1, make sense to me.

File: js/tests/unit/multi-display.test.js (hunk 4 above)
Comment (adglkh) [reply js/tests/unit/multi-display.test.js]: Indeed, thanks for pointing this out.

File: js/anbox-stream-sdk.js (hunk 5 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: +1, yes,this is a leftover from when we agreed to keep `targetElement` for backward compatibility only.
Just dropped that.

File: js/anbox-stream-sdk.js (hunk 6 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: +1, Done.
Thanks for pointing this out

File: js/anbox-stream-sdk.js (hunk 7 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Dropped that due to the leftover after the 1st iteration.

File: js/anbox-stream-sdk.js (hunk 8 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: +1, Fixed that.

File: js/anbox-stream-sdk.js (hunk 9 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Agreed. That shouldn't be hardcoded and it's better to let users customize it.

File: js/anbox-stream-sdk.js (hunk 10 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js:1960]: Thanks for the suggestion!

File: js/anbox-stream-sdk.js (hunk 13 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Sure. fixed that.

File: js/anbox-stream-sdk.js (hunk 14 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: Okay, this is about `Notify AnboxStream to wire a new video element for a display` as it internally 
AnboxWebRTCManager::_onRtcTrack when new video track is available 
|
AnboxWebRTCManager::_onExtraVideoTrack
|
AnboxStream::_onExtraVideoTrack
|
AnboxStream::_onExtraVideoTrack
|
AnboxStream::._options.callbacks.videoTrackAdded(trackIndex);
|
So the consumer(e.g. dashboard) could call AnboxStrae::attachDisplay to attach the display to a video container once the signal is fired.

I just polished that a bit.
Please take another look.

File: js/anbox-stream-sdk.js (hunk 11 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js:1874]: It's a private API exclusively used in _onPointerEvent, where displayId defaults to 0 anyway.
So it should be fine.

File: js/anbox-stream-sdk.js (hunk 12 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js:1998]: See above comment.

**Review by tarek-hanna — APPROVED** [review]
Thanks Gary 👍  LGTM

File: js/anbox-stream-sdk.js (hunk 14 above)
Comment (lorumic) [reply js/anbox-stream-sdk.js]: Yes, it was that "External displays" that felt out of place and left me puzzled. Without it, it's clear. :+1:

File: js/anbox-stream-sdk.js (hunk 9 above)
Comment (lorumic) [reply js/anbox-stream-sdk.js]: @adglkh We still have a fixed black background set at line 1212.

**Review by lorumic — CHANGES_REQUESTED** [review]
Thanks for addressing the comments and fixing the open issues! There is only one leftover black bg colour to be addressed (see [open comment above](https://github.com/canonical/anbox-streaming-sdk/pull/219#discussion_r3611561006)), then we should be good to merge 👍

**Comment by adglkh** [comment]
> Thanks for addressing the comments and fixing the open issues! There is only one leftover black bg colour to be addressed (see [open comment above](https://github.com/canonical/anbox-streaming-sdk/pull/219#discussion_r3611561006)), then we should be good to merge 👍

A: oh, I missed that out. I just fixed that.
    Thanks for pointing this out.

File: js/anbox-stream-sdk.js (hunk 9 above)
Comment (adglkh) [reply js/anbox-stream-sdk.js]: I missed that out and just fixed that.
Please take another look

**Review by lorumic — APPROVED** [review]
Perfect, LGTM 👍
